uat/driver/utils.py

```diff
@@ -0,0 +1,81 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+import subprocess
+
+import tenacity
+from lightkube import Client
+from lightkube.resources.batch_v1 import Job
+from lightkube.resources.core_v1 import Namespace
+
+log = logging.getLogger(__name__)
+
+
+@tenacity.retry(
+    wait=tenacity.wait_exponential(multiplier=2, min=1, max=10),
+    stop=tenacity.stop_after_attempt(30),
+    reraise=True,
+)
+def assert_namespace_active(
+    client: Client,
+    namespace: str,
+):
+    """Test that the provided namespace is Active.
+
+    Retries multiple times to allow for the K8s namespace to be created and reach Active status.
+    """
+    # raises a 404 ApiError if the namespace doesn't exist
+    ns = client.get(Namespace, namespace)
+    phase = ns.status.phase
+
+    log.info(f"Waiting for namespace {namespace} to become 'Active': phase == {phase}")
+    assert phase == "Active", f"Waited too long for namespace {namespace}!"
+
+
+def _log_before_sleep(retry_state):
+    """Custom callback to log the number of seconds before the next attempt."""
+    next_attempt = retry_state.attempt_number
+    delay = retry_state.next_action.sleep
+    log.info(f"Retrying in {int(delay)} seconds (attempts: {next_attempt})")
+
+
+@tenacity.retry(
+    wait=tenacity.wait_exponential(multiplier=2, min=1, max=64),
```

Comment:
subjective, minor suggestion - reject if you disagree
```suggestion
    wait=tenacity.wait_exponential(multiplier=2, min=1, max=32),
```

Reply:
it just feels a bit long for something that is cheap like this wait_for_job.  I'd even cap it at 10s

Reply:
Oh yeah, this is totally subjective! Did you try it with `32`, does it look better? My only concern was not to clutter the pytest output with "waiting" logs (initially I wanted to add these to the debug instead of the info output, but boy is that cluttered).

I'm fine switching this to 32, given that I want to work on streaming the logs anyway (which will deprecate this).

Reply:
I did not try it specifically with 32.  I say go with your gut and someone can change it if they hate us later :)

Reply:
I updated it to 32, this is trivial anyway.